Common/cpp/SharedItems/Shareables.cpp

```diff
@@ -4,41 +4,95 @@ using namespace facebook;
 
 namespace reanimated {
 
-CoreFunction::CoreFunction(
-    JSRuntimeHelper *runtimeHelper,
-    const jsi::Value &workletValue)
-    : runtimeHelper_(runtimeHelper) {
-  jsi::Runtime &rt = *runtimeHelper->rnRuntime();
-  auto workletObject = workletValue.asObject(rt);
-  rnFunction_ = std::make_unique<jsi::Function>(workletObject.asFunction(rt));
-  functionBody_ = workletObject.getPropertyAsObject(rt, "__initData")
-                      .getProperty(rt, "code")
-                      .asString(rt)
-                      .utf8(rt);
-  location_ = "worklet_" +
-      std::to_string(static_cast<uint64_t>(
-          workletObject.getProperty(rt, "__workletHash").getNumber()));
+jsi::Function getValueUnpacker(jsi::Runtime &rt) {
+  auto valueUnpacker = rt.global().getProperty(rt, "__valueUnpacker");
+  assert(valueUnpacker.isObject() && "valueUnpacker not found");
+  return valueUnpacker.asObject(rt).asFunction(rt);
 }
 
-std::unique_ptr<jsi::Function> &CoreFunction::getFunction(jsi::Runtime &rt) {
-  if (runtimeHelper_->isUIRuntime(rt)) {
-    if (uiFunction_ == nullptr) {
-      // maybe need to initialize UI Function
-      // the newline before closing paren is needed because the last line can be
-      // an inline comment (specifically this happens when we attach source maps
-      // at the end) in which case the paren won't be parsed
-      auto codeBuffer = std::make_shared<const jsi::StringBuffer>(
-          "(" + functionBody_ + "\n)");
-      uiFunction_ = std::make_unique<jsi::Function>(
-          rt.evaluateJavaScript(codeBuffer, location_)
-              .asObject(rt)
-              .asFunction(rt));
+#ifdef DEBUG
+
+static const auto callGuardLambda = [](facebook::jsi::Runtime &rt,
+                                       const facebook::jsi::Value &thisVal,
+                                       const facebook::jsi::Value *args,
+                                       size_t count) {
+  return args[0].asObject(rt).asFunction(rt).call(rt, args + 1, count - 1);
+};
+
+jsi::Function getCallGuard(jsi::Runtime &rt) {
+  auto callGuard = rt.global().getProperty(rt, "__callGuardDEV");
+  if (callGuard.isObject()) {
+    // use JS-based implementation if available
+    return callGuard.asObject(rt).asFunction(rt);
+  }
+  // fallback to C++ JSI implementation
+  // this is necessary to install __callGuardDEV itself
```

Comment:
We use the C++ JSI-based implementation only once when we run a worklet that registers `callGuardDEV` itself. Each consecutive call of `runOnRuntimeGuarded` will use the JS-based `callGuardDEV` since it will have been already installed.